File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    terms = {k.split(' ')[0]:int(k.split(' ')[1]) for k in terms}
    ty, tm, td = map(int, today.split('.'))
    cnt = 1
    answer = []
    for privacy in privacies:
        begin, term = privacy.split(' ')
        
        by, bm, bd = map(int, begin.split('.'))

        term = terms[term]
        d = term * 28 - 1
        nd = (bd + d) % 28 if (bd + d) % 28 else 28
        nm = (((bd + d) // 28 ) + bm) % 12 if (((bd + d) // 28 ) + bm) % 12 else 12
        ny = ((((bd + d) // 28 ) + bm) // 12) + by
        if nm == 12: ny -= 1
        if nd == 28: nm -= 1
        
        
        print(ny, nm, nd)
        
        if ty - ny > 0:
            answer.append(cnt)
            cnt +=1
            continue
        
        if ty - ny == 0:
            if tm - nm > 0:
                answer.append(cnt)
                cnt +=1
                continue
            elif tm - nm == 0:
                if td - nd > 0:
                    answer.append(cnt)
                    cnt += 1
                    continue
        cnt += 1

    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.go
import (
	"strconv"
	"strings"
)

func solution(today string, terms []string, privacies []string) []int {
	var answer []int

	todayDate := toDate(today)

	termMap := make(map[string]int)

	for _, term := range terms {
		termParts := strings.Split(term, " ")
		m, _ := strconv.Atoi(termParts[1])
		termMap[termParts[0]] = m
	}

	for i, privacy := range privacies {
		datePrivacy := strings.Split(privacy, " ")
		addM := termMap[datePrivacy[1]]
		targetDate := calcDate(toDate(datePrivacy[0]), addM)
		if isTimeout(targetDate, todayDate) {
			answer = append(answer, i+1)
		}
	}

	return answer
}

type Date struct {
	Y int
	M int
	D int
}

func toDate(stringDate string) Date {
	parts := strings.Split(stringDate, ".")
	Y, _ := strconv.Atoi(parts[0])
	M, _ := strconv.Atoi(parts[1])
	D, _ := strconv.Atoi(parts[2])
	return Date{Y: Y, M: M, D: D}
}

func calcDate(date Date, m int) Date {
	date.M += m
	ay := 0

	for date.M > 12 {
		date.M -= 12
		ay += 1
	}
	date.Y += ay

	return date
}

func isTimeout(date Date, now Date) bool {
	intDate := date.Y*10000 + date.M*100 + date.D
	intNow := now.Y*10000 + now.M*100 + now.D
	if intDate > intNow {
		return false
	}
	return true
}
